crates/federate/src/worker.rs

```diff
@@ -1,132 +1,192 @@
+      in_flight: 0,
     };
+
     worker.loop_until_stopped().await
   }
   /// loop fetch new activities from db and send them to the inboxes of the given instances
   /// this worker only returns if (a) there is an internal error or (b) the cancellation token is
   /// cancelled (graceful exit)
-  pub(crate) async fn loop_until_stopped(&mut self) -> Result<(), anyhow::Error> {
-    debug!("Starting federation worker for {}", self.instance.domain);
-    let save_state_every = chrono::Duration::from_std(SAVE_STATE_EVERY_TIME).expect("not negative");
-
-    self.update_communities().await?;
+  async fn loop_until_stopped(&mut self) -> Result<()> {
     self.initial_fail_sleep().await?;
+    let (mut last_sent_id, mut newest_id) = self.get_latest_ids().await?;
+
     while !self.stop.is_cancelled() {
-      self.loop_batch().await?;
-      if self.stop.is_cancelled() {
-        break;
+      // check if we need to wait for a send to finish before sending the next one
+      // we wait if (a) the last request failed, only if a request is already in flight (not at the
+      // start of the loop) or (b) if we have too many successfuls in memory or (c) if we have
+      // too many in flight
+      let need_wait_for_event = (self.in_flight != 0 && self.state.fail_count > 0)
+        || self.successfuls.len() >= MAX_SUCCESSFULS
+        || self.in_flight >= self.federation_worker_config.concurrent_sends_per_instance;
+      if need_wait_for_event || self.receive_send_result.len() > MIN_ACTIVITY_SEND_RESULTS_TO_HANDLE
+      {
+        // if len() > 0 then this does not block and allows us to write to db more often
+        // if len is 0 then this means we wait for something to change our above conditions,
+        // which can only happen by an event sent into the channel
+        self.handle_send_results().await?;
+        // handle_send_results does not guarantee that we are now in a condition where we want to
+        // send a new one, so repeat this check until the if no longer applies
+        continue;
       }
-      if (Utc::now() - self.last_state_insert) > save_state_every {
-        self.save_and_send_state().await?;
+
+      // send a new activity if there is one
+      self.inbox_collector.update_communities().await?;
+      let next_id_to_send = ActivityId(last_sent_id.0 + 1);
+      {
+        // sanity check: calculate next id to send based on the last id and the in flight requests
+        let expected_next_id = self.state.last_successful_id.map(|last_successful_id| {
+          last_successful_id.0 + (self.successfuls.len() as i64) + self.in_flight + 1
+        });
+        // compare to next id based on incrementing
+        if expected_next_id != Some(next_id_to_send.0) {
+          anyhow::bail!(
+            "{}: next id to send is not as expected: {:?} != {:?}",
+            self.instance.domain,
+            expected_next_id,
+            next_id_to_send
+          )
+        }
```

Comment:
So if these values are inconsistent it wont send anything. Do you expect this to fix itself in later runs? Otherwise it could break federation for a longer time.

Reply:
the bail kills the instance worker, which is then recreated by the sendmanager. the database only stores a single value which makes this inconsistency impossible, so if this happens it should fix itself with max damage being a few dupe sends (same as restarting the process would)